### PR TITLE
Fix/separate migration bins

### PR DESCRIPTION
**Pull request by gammazero**
### Fetch Individual Migrations as Needed for Update 
This PR provides new logic to fetch individual migrations from the IPFS distribution site, instead of fetching a monolithic binary that contains all migrations.  This is in support of future migrations as described in [fs-repo-migrations/issues/98](https://github.com/ipfs/fs-repo-migrations/issues/98). Each migration is available as an individual distribution so that only the migrations needed for an upgrade are fetched from the distribution site and unpacked.  Logic is included to execute migrations in forwards or reverse order, for upgrade and revert respectively. 

### Consolidate Distribution Fetch Logic
This PR consolidates logic for fetching files from the IPFS distribution site that was partially duplicated in [ipfs/fs-repo-migrations](https://github.com/ipfs/fs-repo-migrations) and in [ipfs/ipfs-update](https://github.com/ipfs/ipfs-update).  Those projects now import their distribution fetch logic from this library.

This PR relies on the [distributions PR 327](https://github.com/ipfs/distributions/pull/327)

### CODE REVIEW

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [inline repo/fsrepo/migrations/fetch.go:23]: This is a security issue and should be changed to use an IPFS path.
```go
ipfsDist = "/ipfs/Qme8pJhBidEUXRdpcWLGR2fkG5kdwVnaMh3kabjfP8zz7Y"
```
The downside to this is that nothing newer than what is in that distribution can be fetched.  In the case of `ipfs-update` that means either downloading the latest `ipfs-update`, or specifying the distribution to fetch from (using `IPFS_DIST_PATH` environ var or `-distpath` option).

Comment (aschmahmann) [reply repo/fsrepo/migrations/fetch.go:23]: I think for now we should keep existing behavior:
- `ipfs-update` pulls from the DNSLink record
- `go-ipfs` repo migrations pull from the version of dist bundled with the binary

I'd like to have an option for automatic updating of go-ipfs, but that's a separate issue + PR.

The short reasoning for why we're using different semantics for `ipfs-update` vs `go-ipfs` migrations is basically that go-ipfs migrations happen automatically whereas ipfs-update is something a user manually runs where the alternative is that they'd likely go to GitHub or dist.ipfs.io anyway.

Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:23]: OK, I will have go-ipfs use the ipfs path and ipfs-update use a DNSLink record. What you prefer the default behavior of this library to be, ipfs or dnslink?

Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:26]: Done.

File: repo/fsrepo/migrations/fetch.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/fetch.go:75]: The current setup here is that we have a global function `FetchBinary` that internally calls `fetchHttp` and `fetchIpfs`, where `fetchIpfs` is using go-ipfs-api.

This is a bit unfortunate as it:
1. Adds a dependency on go-ipfs-api
2. Doesn't really leave us room for figuring out a way for go-ipfs to download the binary over IPFS without reworking things (i.e. when go-ipfs automatically does a migration it's not going to be able to use IPFS to handle this since the node isn't and can't be running yet)

One way we might clean this up is to have a `Fetcher` interface that just gets the archive and is implemented by an `HttpFetcher` and an `IpfsFetcher` and we can pass the Fetcher to the `FetchBinary` function. Additionally, we can have the `IpfsFetcher` take a CoreAPI and then callers like ipfs-update can use https://github.com/ipfs/go-ipfs-http-client as their CoreAPI implementation.

WDYT?

File: repo/fsrepo/fsrepo.go
Comment (aschmahmann) [inline repo/fsrepo/fsrepo.go:145]: Why did we change this pattern, just less indirection?

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Left a number of comments, questions and suggestions.

One of the recurrent things is that the tests seem to assume we're online which doesn't seem like a great idea for Go tests which are supposed to be more like unit tests than integration tests. We can go into it more on the threaded comments though.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:46]: It was intended to simplify APIs.  Multiple APIs need to know the location of the distribution site, this allows the location to be set once, if needed, instead of passing another value into multiple APIs.  I also assume this was why it was previously a global variable, and not a function argument: https://github.com/ipfs/go-ipfs/blob/master/repo/fsrepo/migrations/migrations.go#L18

Would you prefer to see the distribution location passed into all the APIs that need it?

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:103]: It looks like that is handled immediately above, in the code:
```go
if !fi.IsDir() {
    return "", &os.PathError{
        Op:   "FetchBinary",
        Path: out,
        Err:  os.ErrExist,
    }
}
```

File: repo/fsrepo/fsrepo.go
Comment (gammazero) [reply repo/fsrepo/fsrepo.go:145]: There were multiple implementations of this same code, here and in ipfs-update.  I chose the API that I felt was the simplest and most readable.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:269]: Looks like `rootDir == name` is always true on our distribution site.  I wanted to allow for an archive to have a name (prefix) that differed from the distribution name, but that is never happen unless someone is using a custom distribution site.  So, the `arcName` argument can be removed from `FetchBinary`

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:75]: I see your point regarding 1 and 2.  The Ipfs fetcher was to support `ipfs-update` and `fs-repo-migrations`, and I did not consider that for automated migrations.

As far as passing an `IpfsFetcher` or `HttpFetcher` to `FetchBinary`, that means that a decision has already been made to use one or the other.  With `ipfs-update`, I wanted to preserve its ability to try fetching from IPFS, and if not available, then fail-over to fetching from HTTP.  That could be solved by a "compountFetcher" implementation whose `Fetch` method tries one, then the other. 

The other issue with passing in a Fetcher is that I would also need to pass the Fetcher to every API that needs to fetch anything.  This includes fetching available and latest versions and running migrations.  Is it reasonable for all these APIs to require a Fetcher?

File: repo/fsrepo/migrations/fetch_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch_test.go:47]: Agree on not depending on net connectivity; should set up a dummy [httptest](https://golang.org/pkg/net/http/httptest/) server.  This is testing that something can be retrieved over HTTP and that an error is handled.

File: repo/fsrepo/migrations/migrations.go
Comment (gammazero) [reply repo/fsrepo/migrations/migrations.go:24]: The code that calls this already does that.  This is only called once a final decision has been made to run the migration.

File: repo/fsrepo/migrations/migrations.go
Comment (gammazero) [reply repo/fsrepo/migrations/migrations.go:38]: Mostly, I wanted this diagnostic/progress output sent to stderr to prevent it from hindering any future conventional output that could serve as input to another program.  I also like having timestamps so that I can see how long individual items take to download, so I switched to using `log`.  Please let me know what you prefer.

This is probably part of a larger conversation about standardizing program output... do we user `log`, `stump`, `fmt`, etc, something else, and what output goes to what stream.

File: repo/fsrepo/migrations/fetch.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/fetch.go:103]: Let's say out is `/fooDir` and the binary name is `myBin`, and there already exists a file or folder named `/fooDir/myBin`.

1. Check if `/fooDir` exists, it does.
2. Check if `/fooDir` is a directory, it is.
3. Set `out=/fooDir/myBin`
4. Try writing to `/fooDir/myBin` and fail. This fail will happen later and AFAICT the point of this check is to fail before we do the download

File: repo/fsrepo/migrations/fetch.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/fetch.go:269]: yep, simplicity seems good. If we ever have a need for this we'd have to call this function with the new parameters anyway and we could just make a PR to change things.

File: repo/fsrepo/migrations/migrations.go
Comment (gammazero) [reply repo/fsrepo/migrations/migrations.go:188]: Yes, this is completely safe because:
- Slice is never resized
- Each goroutine has a separate element to write to; same as if each wrote to separate closed-over variable
- Slice is only read after passing `wg.Wait()` synchronization point

This is a safe and efficient implementation.  It is useful when there is a separate place for each goroutine to write its output, and there is nothing for the reader to do until all goroutines finish.  

The tl;dr version:
If multiple goroutines access a variable concurrently, and at least one of the accesses is a write, then synchronization is required.  Here the rule is not violated, because the goroutines do not write the slice value (slice header), they only read it when indexing the slice. Each goroutine only modifies a single, different, designated slice element. Since each slice element has its own address (own memory space), they are like distinct variables. [Spec: Variables](https://golang.org/ref/spec#Variables) says,
>Structured variables of [array](https://golang.org/ref/spec#Array_types), [slice](https://golang.org/ref/spec#Slice_types), and [struct](https://golang.org/ref/spec#Struct_types) types have elements and fields that may be [addressed](https://golang.org/ref/spec#Address_operators) individually. Each such element acts like a variable.

Follwoing the rule, results cannot be read without synchronization.  The `waitgroup` provides sufficient synchronization because it ensures all goroutines complete and cannot modify elements after `wg.Wait()` returns, and ensures the modification are visible to the goroutine that called `wg.Wait()`

File: repo/fsrepo/migrations/migrations.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/migrations.go:24]: I think that'd be better. For example, we call RunMigration in the daemon init as `err = migrations.RunMigration(cctx.Context(), fsrepo.RepoVersion, "")` and no determination around reversion has been made there.

File: repo/fsrepo/migrations/fetch.go
Comment (aschmahmann) [reply repo/fsrepo/migrations/fetch.go:75]: > compoundFetcher

Yep, exactly my thinking.

>  Is it reasonable for all these APIs to require a Fetcher?

I think so, while we could use globals I'm generally a fan of explicitness even if it adds a bit of verbosity to the API.

File: repo/fsrepo/migrations/unpack.go
Comment (gammazero) [reply repo/fsrepo/migrations/unpack.go:53]: Yep, good call.  Might want to expose the underlying error to callers.  There are probably many places where this should be done.

File: repo/fsrepo/migrations/unpack.go
Comment (gammazero) [reply repo/fsrepo/migrations/unpack.go:86]: Yes

File: repo/fsrepo/fsrepo.go
Comment (aschmahmann) [reply repo/fsrepo/fsrepo.go:145]: Fair enough, I mostly wanted to know why we did the change. This way does a little more repeated work in exchange for a slightly cleaner API which SGTM.

File: repo/fsrepo/migrations/versions.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions.go:31]: Not for testing.  This was already present in the code from ipfs-update [here](https://github.com/ipfs/ipfs-update/blob/master/lib/versions.go#L85), so I did not want to remove it in case there was some situation in which it was needed.

Keep it? Leave it?

File: repo/fsrepo/migrations/unpack.go
Comment (gammazero) [reply repo/fsrepo/migrations/unpack.go:100]: You are not missing anything, and this should be changed.  It was left over from ipfs-update [here](https://github.com/ipfs/ipfs-update/blob/master/lib/archive.go#L71)

From the perspective of applications unpacking an archive, that generally (always) write the extracted binary to a temp location, the message is OK.  No OK from the perspective of the library.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:46]: This is now passed to the fetcher(s) and is not a global.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go]: Fetcher is now passed into functions that need to fetch from network. ipfs-update will provide IpfsFetcher implementation of Fetcher interface.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:56]: Fixed.

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go]: Done.

File: repo/fsrepo/migrations/fetch_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch_test.go]: Tests use a httptest server now, and fetchers are pointed at it.  The test server can serve content to simulate the real thing.

File: repo/fsrepo/migrations/fetch_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch_test.go]: Fixed

File: repo/fsrepo/migrations/migrations.go
Comment (gammazero) [reply repo/fsrepo/migrations/migrations.go]: Done.

File: repo/fsrepo/migrations/unpack.go
Comment (gammazero) [reply repo/fsrepo/migrations/unpack.go]: Done

File: repo/fsrepo/migrations/unpack.go
Comment (gammazero) [reply repo/fsrepo/migrations/unpack.go]: Done.

File: repo/fsrepo/migrations/versions.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions.go]: Done

File: repo/fsrepo/migrations/versions.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions.go]: It must have been there for some reason at some time... will leave it for now.

File: repo/fsrepo/migrations/ipfsdir.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsdir.go:38]: @Stebalien, this seems fine to me. Is it reasonable to ignore the reasoning behind the limiting here? https://github.com/ipfs/go-ipfs/blob/3393b4a39efae379daed421ec1754499affe849f/repo/fsrepo/fsrepo.go#L338

File: repo/fsrepo/migrations/ipfsdir.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsdir.go:31]: @gammazero this is moving out right since go-ipfs-api does this anyway and that's getting called by the relevant binaries (ipfs-update + fs-repo-migrations)?

File: repo/fsrepo/migrations/ipfsdir.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/ipfsdir.go:49]: This code is currently in ipfs-update. It looks like it's probably fine although it seems a bit brittle. @Stebalien do we only really support `/ipX/address/tcp/port` in the API file, or at least only support that for direct HTTP usage?

Perhaps switching over to go-ipfs-http-client would put us in a better spot here https://github.com/ipfs/go-ipfs-http-client/blob/3af36210f80fb86aae50da582b494ceddd64c3de/api.go#L99.

----

@gammazero if we do then it's reasonable to push the switch from go-ipfs-api to go-ipfs-http-client into a separate PR since it's unrelated to the migrations work

File: repo/fsrepo/migrations/versions.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/versions.go:81]: What is this used for, it doesn't seemed to be called anywhere?

Also, using exec like this is generally a bad idea since they user could have renamed their binary (e.g. ipfs-update will automatically at the version number to the end of binaries that are fetched). We could read the version corresponding to this package from the Go code, or if we wanted to read the repo version we could read it from the file itself (which would require passing the ipfs repo folder in too).

File: repo/fsrepo/migrations/versions_test.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/versions_test.go:43]: was removing these lines accidental?

File: repo/fsrepo/migrations/versions_test.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/versions_test.go:60]: again, bad idea to use exec but also not sure why this function (+ test are needed)

**Review by aschmahmann — CHANGES_REQUESTED** [review]

File: repo/fsrepo/migrations/fetch.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetch.go:66]: Intentional.  I think this is the most appropriate error to return.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go:30]: Correct.  Eventually it may be used by go-ipfs also, to implement an IPFS fetcher in addition to the http fetcher.

File: repo/fsrepo/migrations/httpfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/httpfetcher.go:71]: I would prefer to leave it blank.  If this is useful for access logs at the distribution site, then maybe this can be an option settable in the HTTP fetcher.

File: repo/fsrepo/migrations/ipfsdir.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsdir.go:31]: Yes.  That should have been moved out.

File: repo/fsrepo/migrations/versions.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions.go:81]: I will move the exec portion back into ipfs-update to be used if IPFS_PATH is not set.

File: repo/fsrepo/migrations/versions_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions_test.go:60]: moving back to ipfs-update

File: repo/fsrepo/migrations/httpfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/httpfetcher.go:33]: The idea was that these are things that only need to be set in exceptional cases. However, it does invite mutating the struct in the middle of operations, so not a good pattern for initialization.  Changing to using constructor arguments.

File: repo/fsrepo/migrations/httpfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/httpfetcher.go]: Done.  There is a constructor argument now.

File: repo/fsrepo/migrations/ipfsdir.go
Comment (gammazero) [reply repo/fsrepo/migrations/ipfsdir.go]: Done.

File: repo/fsrepo/migrations/versions.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions.go]: Done.

File: repo/fsrepo/migrations/versions_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions_test.go]: Yes, fixed.

File: repo/fsrepo/migrations/versions_test.go
Comment (gammazero) [reply repo/fsrepo/migrations/versions_test.go]: Done.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go]: Done. Specify using constructor argument instead.

File: repo/fsrepo/migrations/httpfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/httpfetcher.go]: Done. Using constructor arguments now.

File: repo/fsrepo/migrations/httpfetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/httpfetcher.go:33]: This is less lines of code than the functional options approach, although it comes at the cost of needing to break the function definition if we add any more options. Also, by returning `*HttpFetcher` instead of `(*HttpFetcher, error)` we pass the error checking into the call to `Fetch`.

I like the options pattern + errors despite the verbosity because you almost never have to break the interface. However, given that this likely to be used mostly by go-ipfs, ipfs-update, and fs-repo-migrations and the struct is simple enough a consumer could probably just create their own maybe simple, but breakable, is ok here.

File: repo/fsrepo/migrations/fetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/fetcher.go:11]: TODO: Update before merging PR (this is out of date anyhow, but also we need to publish to dist with the new migration binaries anyway)

File: repo/fsrepo/migrations/fetcher.go
Comment (aschmahmann) [inline repo/fsrepo/migrations/fetcher.go:13]: Calling this IPNS when it's really DNSLink is pretty unfortunate, especially since things are already confusing given the `/ipns` prefix. How about referencing DNSLink or just using "Latest" like?

```suggestion
	// Latest distribution path.  Default for fetchers.
	LatestIpfsDist = "/ipns/dist.ipfs.io"
```'

**Review by aschmahmann — COMMENTED** [review]
Left a few small comments, but LGTM.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go:11]: Before merge, I will create a new dist with the latest migration binaries and change all references to this.

File: repo/fsrepo/migrations/httpfetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/httpfetcher.go:33]: I think the simpler approach is OK in this case, so will leave it as is.

File: repo/fsrepo/migrations/fetcher.go
Comment (gammazero) [reply repo/fsrepo/migrations/fetcher.go]: Done.

**Pull request merged**
